src/main/java/shef/servlets/UserServlet.java

```diff
@@ -92,36 +93,62 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     UserService userService = UserServiceFactory.getUserService();
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-
+      
+    // These properties don't come from the request, so we get them here.
     String id = userService.getCurrentUser().getUserId();
     String email = userService.getCurrentUser().getEmail();
-    String username = request.getParameter("username-input");
-    String location = request.getParameter("location-input");
-    String profilePictureUrl = getUploadedFileUrl(request, "profile-pic-upload");
-    String bio = request.getParameter("bio-input");
+    String profilePicUrl = getUploadedFileUrl(request, "profile-pic-url");
+
+    // Get the properties from the request.
+    Map<String, String[]> parameterMap = request.getParameterMap();
+    Set entrySet = parameterMap.entrySet();
+    Iterator it = entrySet.iterator();
 
     Key userKey = KeyFactory.createKey("User", id);
+    String keyString = KeyFactory.keyToString(userKey);
 
-    if (profilePictureUrl == null) {
-      try {
-        Entity originalEntity = datastore.get(userKey);
-        profilePictureUrl = (String) originalEntity.getProperty("profile-picture-url");
-      } catch (EntityNotFoundException e) {
-        throw new IOException("No profile picture provided.");
+    try {
+      Entity user = datastore.get(userKey);
+      // If updating an existing user, just update the changed fields.
+      while(it.hasNext()) {
+        Map.Entry<String,String[]> entry = (Map.Entry<String,String[]>) it.next();
+        String key = entry.getKey();
+        String[] value = entry.getValue();
+
+        if(value != null) {
+          user.setProperty(key, value[0]);
+        }
       }
-    }
 
-    // Create a new User entity with data from the request.
-    Entity userEntity = new Entity(userKey);
-    userEntity.setProperty("email", email);
-    userEntity.setProperty("username", username);
-    userEntity.setProperty("location", location);
-    userEntity.setProperty("profile-picture-url", profilePictureUrl);
-    userEntity.setProperty("bio", bio);
-
-    // Store the User entity in Datastore.
-    datastore.put(userEntity);
-    response.sendRedirect("/user-list-test.html");
+      // These properties don't come from the request, so we add them here.
+      user.setProperty("email", email);
+      if(profilePicUrl != null) {
+        user.setProperty("profile-pic-url", profilePicUrl);
+      }
+
+      // Store the User entity in Datastore.
+      datastore.put(user);
+      response.sendRedirect("/profile-page.html?key=" + keyString);
+    } catch (EntityNotFoundException e) {
+      // Create a new User entity with data from the request.
+      Entity user = new Entity(userKey);
+      while(it.hasNext()) {
```

Comment:
There is a lot of duplicate code within try and catch blocks. You can declare the user variable outside these blocks, assign within them based on the situation and then access it after the blocks.Is this feasible?

Reply:
Excellent suggestion, I was able to pretty much get rid of the duplicate code.